src/unsigned_stubs.c

```diff
@@ -37,6 +49,56 @@
                                  OP Uint_custom_val(SIZE, b));             \
   }
 
+#define UINT_OF_STRING(BITS, COPY)                                           \
+  value integers_uint ## BITS ## _of_string(value a)                         \
+  {                                                                          \
+    TYPE(BITS) u, max_prefix;                                                \
+    const char *pos = String_val(a);                                         \
+    int base = 10, d;                                                        \
+                                                                             \
+    /* Strip a leading + sign, if given */                                   \
+    if (*pos == '+') pos++;                                                  \
+    if (*pos == '0') {                                                       \
+      switch (pos[1]) {                                                      \
+        case 'x': case 'X':                                                  \
+          base = 16; pos += 2; break;                                        \
+        case 'o': case 'O':                                                  \
+          base = 8; pos += 2; break;                                         \
+        case 'b': case 'B':                                                  \
+          base = 2; pos += 2; break;                                         \
+        case 'u': case 'U': /* Unsigned prefix. No-op for unsigned types */  \
+          pos += 2; break;                                                   \
+      }                                                                      \
+    }                                                                        \
+                                                                             \
+    max_prefix = ((TYPE(BITS)) 1 << (BITS - 1)) / base;                      \
```

Comment:
I don't follow this `max_prefix` logic.  When `BITS` is `8` and `base` is 10, we have
```C
max_prefix = ((uint8_t)1 << 7) / 10
```
which is `128 / 10` i.e. `12`, and won't allow parsing a `uint8_t` of 130 or above.  Perhaps something like `(((TYPE(BITS)-1) - base) / base) + 1` is sufficient.

Reply:
Sorry, I ought to have added more description here. The goal of that expression is, 'find the largest number which, if we've already parsed it, parsing another digit will cause us to overflow regardless of what it is' (ie. `x s.t. x < size(TYPE(BITS)) and base * x > size(TYPE(BITS))`). This also leans on the fact that all of the bases we support are all even, and thus not divisors of `-1` for any of the types, otherwise the second inequality may not be strict.

Your expression also works, but ends up as 3 instructions rather than 1, because base is a variable (so we sub, divide and then add rather than just divide). I'm happy to add some docs; would you prefer if I also changed the expression too?

Reply:
I don't think it's correct as it stands:
```ocaml
# UInt8.of_string "129";;
- : Unsigned.UInt8.t = 129
# UInt8.of_string "130";;
Exception: Failure "int_of_string".
```

Reply:
Sorry, yes. I'd misunderstood your comment :man_facepalming: It should work now (essentially what you had, but leveraging the 'base doesn't divide -1' property).